File: 43.multiply-strings.java
/*
 * @lc app=leetcode id=43 lang=java
 *
 * [43] Multiply Strings
 */

// @lc code=start

class Solution {
    public String multiply(String num1, String num2) {
        // for each digits in nums, do the mul with num2
        // 1. from right to left, num1[i]*num2[j]=mul={cin, cout}
        // 2. out[i+j]+=cin, out[i+j-1]+=cout;
        int m = num1.length(), n = num2.length();
        int[] out = new int[m + n];

        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                int mul = (num1.charAt(i) - '0') * (num2.charAt(j) - '0');
                int p1 = i + j, p2 = i + j + 1;
                int sum = mul + out[p2];

                out[p1] += sum / 10;
                out[p2] = (sum) % 10;
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int p : out)
            // no leading zeros
            if (sb.length() != 0 || p != 0)
                sb.append(p);

        return sb.length() == 0 ? "0" : sb.toString();
    }
}
// @lc code=end

File: 43.multiply-strings.py
#
# @lc app=leetcode id=43 lang=python3
#
# [43] Multiply Strings
#

# @lc code=start
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        m,n=len(num1),len(num2)
        res=[0]*(m+n)
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                digit1=ord(num1[i])-ord('0')
                digit2=ord(num2[j])-ord('0')
                res[i+j+1]+=digit1*digit2

        for x in range(m+n-1,0,-1):
            carry=res[x]//10
            res[x]%=10
            res[x-1]+=carry

        idx=0
        while idx<m+n and res[idx]==0:
            idx+=1

        ans=''.join(str(digit) for digit in res[idx:]) 
        return ans or "0"         
# @lc code=end

